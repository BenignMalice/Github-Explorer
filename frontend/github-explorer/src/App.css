/* Reset default styles and set a background color */
body {
  font-family: "Arial", sans-serif;
  margin: 0;
  padding: 0;
  background-color: #f4f4f4;
}
/* Container styles for the main content */
.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  max-width: 800px;
  margin: 0 auto;
  padding: 20px;
  background-color: #fff;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
  border-radius: 8px;
  margin-top: 30px;
}
/* Heading styles */
h1 {
  color: #333;
}
/* Input styles */
input {
  padding: 10px;
  font-size: 16px;
  margin-right: 10px;
  border: 1px solid #ccc;
  border-radius: 4px;
}
/* Button styles */
button {
  padding: 10px;
  font-size: 16px;
  background-color: #4caf50;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}
/* Button hover effect */
button:hover {
  background-color: #45a049;
}
/* Styles for the user section */
.user-section {
  display: flex;
  align-items: center;
  margin-top: 20px;
  margin-bottom: 20px;
}
/* Avatar styles */
.avatar {
  width: 100px;
  height: 100px;
  border-radius: 50%;
  margin-right: 20px;
}
/* User info styles */
.user-info {
  flex-grow: 1;
}
/* Styles for the list of repositories */
.repositories {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}
/* Repository list styles */
.repository-list {
  display: flex;
  flex-wrap: wrap;
}
/* Individual repository item styles */
.repository-item {
  flex: 0 0 calc(33.33% - 20px);
  margin: 10px;
  padding: 15px;
  background-color: #fff;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
  border-radius: 8px;
  cursor: pointer;
  border: 1px solid #ddd;
  transition: background-color 0.3s;
}
/* Repository item hover effect */
.repository-item:hover {
  background-color: #4caf50;
}
/* Styles for repository details */
.repository-details {
  margin-top: 20px;
  border: 1px solid #ddd;
  padding: 15px;
  border-radius: 8px;
  background-color: #fff;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
}
/* Styles for commit details */
.commit-details {
  margin-top: 20px;
  border: 1px solid #ddd;
  padding: 15px;
  border-radius: 8px;
  background-color: #fff;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
}
/* Error message styles */
.error {
  color: red;
}
