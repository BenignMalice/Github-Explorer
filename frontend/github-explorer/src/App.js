// Import necessary modules
import React, { useState, useEffect } from "react";
import "./App.css";
import { AiOutlineLoading } from "react-icons/ai";

// Define the main App component
function App() {
  // State variables to manage user input, data, errors, and loading state
  const [username, setUsername] = useState("");
  const [userData, setUserData] = useState(null);
  const [error, setError] = useState(null);
  const [repos, setRepos] = useState([]);
  const [selectedRepo, setSelectedRepo] = useState(null);
  const [repoDetails, setRepoDetails] = useState(null);
  const [commits, setCommits] = useState([]);
  const [readmeContent, setReadmeContent] = useState(null);
  const [loading, setLoading] = useState(false);

  // Function to handle user search
  const handleSearch = async () => {
    try {
      setLoading(true);
      // Fetch user data from the Express backend
      const response = await fetch(
        `http://localhost:3001/api/users?names=${encodeURIComponent(
          JSON.stringify([username])
        )}`
      );
      const data = await response.json();

      if (data && !data.error) {
        // If successful, update user data and fetch repositories
        setUserData(data[0]);
        setError(null);

        // Fetch user repositories from the Express backend
        const reposResponse = await fetch(
          `http://localhost:3001/api/users/${username}/repos`
        );

        if (!reposResponse.ok) {
          throw new Error("Unable to fetch repositories");
        }

        const reposData = await reposResponse.json();
        setRepos(reposData);
        setSelectedRepo(null);
        setRepoDetails(null);
        setCommits([]);
      } else {
        // If there is an error, update error state
        setUserData(null);
        setRepos([]);
        setSelectedRepo(null);
        setRepoDetails(null);
        setCommits([]);
        setError(data.error || "Error searching for user");
      }
    } catch (err) {
      // Handle errors and update error state
      setUserData(null);
      setRepos([]);
      setSelectedRepo(null);
      setRepoDetails(null);
      setCommits([]);
      setError(err.message);
    } finally {
      setLoading(false);
    }
  };

  // Function to handle click on a repository
  const handleRepoClick = async (repoName) => {
    try {
      // Fetch repository details from the Express backend
      const repoDetailsResponse = await fetch(
        `http://localhost:3001/api/users/${username}/repos/${repoName}`
      );

      if (!repoDetailsResponse.ok) {
        throw new Error("Unable to fetch repository details");
      }

      const repoDetailsData = await repoDetailsResponse.json();
      setRepoDetails(repoDetailsData);

      // Fetch the last 5 commits
      const commitsResponse = await fetch(
        `https://api.github.com/repos/${username}/${repoName}/commits?per_page=5`
      );

      if (!commitsResponse.ok) {
        throw new Error("Unable to fetch commits");
      }

      const commitsData = await commitsResponse.json();
      setCommits(commitsData);

      // Fetch README.md content
      const readmeResponse = await fetch(
        `https://api.github.com/repos/${username}/${repoName}/readme`
      );

      if (readmeResponse.ok) {
        const readmeData = await readmeResponse.json();

        if (readmeData && readmeData.content) {
          // Decode base64 content
          const decodedContent = atob(readmeData.content);
          // Set the README.md content
          setReadmeContent(decodedContent);
        }
      }
    } catch (err) {
      // Handle errors and update relevant states
      console.error("Error fetching repository details:", err);
      setRepoDetails(null);
      setCommits([]);
      setReadmeContent(null);
    }

    // Set the selected repo for styling purposes
    setSelectedRepo(repoName);
  };

  return (
    <div className="container">
      <h1>Github User Search</h1>
      <div>
        <input
          type="text"
          placeholder="Enter Github username"
          value={username}
          onChange={(e) => setUsername(e.target.value)}
        />
        <button onClick={handleSearch} disabled={loading}>
          {loading ? <AiOutlineLoading className="loading-icon" /> : "Search"}
        </button>
      </div>

      {error && <p className="error">{error}</p>}

      {userData && (
        <div className="user-section">
          <img className="avatar" src={userData.avatar_url} alt="Profile" />
          <div className="user-info">
            <h2>{userData.login}</h2>

            <p>
              GitHub Profile:{" "}
              <a
                href={`https://github.com/${userData.login}`}
                target="_blank"
                rel="noopener noreferrer"
              >
                {`https://github.com/${userData.login}`}
              </a>
            </p>
            <p>{userData.bio}</p>
          </div>
        </div>
      )}

      {repos.length > 0 && (
        <div className="repositories">
          <h3>Repositories:</h3>
          <div className="repository-list">
            {repos.map((repo) => (
              <div
                key={repo.name}
                onClick={() => handleRepoClick(repo.name)}
                className={`repository-item ${
                  selectedRepo === repo.name ? "selected-repo" : ""
                }`}
              >
                {repo.name}
              </div>
            ))}
          </div>
        </div>
      )}

      {repoDetails && (
        <div id="repository-details" className="repository-details">
          <h3>Repository Details:</h3>

          <p>{`Name: ${repoDetails.name}`}</p>
          <p>{`Description: ${repoDetails.description || "N/A"}`}</p>
          <p>{`Created at: ${new Date(
            repoDetails.created_at
          ).toLocaleDateString()}`}</p>
          <p>{`Last commit at: ${new Date(
            repoDetails.updated_at
          ).toLocaleDateString()}`}</p>
          <p>
            Link:{" "}
            <a
              href={`https://www.github.com/${username}/${repoDetails.name}`}
              target="_blank"
              rel="noopener noreferrer"
            >
              {repoDetails.name}
            </a>
          </p>
        </div>
      )}

      {commits.length > 0 && (
        <div className="commit-details">
          <h3>Last 5 Commits:</h3>
          <ul>
            {commits.map((commit) => (
              <li key={commit.sha}>{commit.commit.message}</li>
            ))}
          </ul>
        </div>
      )}
    </div>
  );
}

export default App;
