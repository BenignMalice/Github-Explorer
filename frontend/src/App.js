// Import necessary modules
import React, { useState } from "react";
import SearchForm from "./components/SearchForm";
import UserInfo from "./components/UserInfo";
import Repositories from "./components/Repositories";
import RepositoryDetails from "./components/RepositoryDetails";
import CommitDetails from "./components/CommitDetails";
import "./App.css";

// Define the main App component
function App() {
  // State variables to manage user input, data, errors, and loading state
  const [username, setUsername] = useState("");
  const [userData, setUserData] = useState(null);
  const [error, setError] = useState(null);
  const [repos, setRepos] = useState([]);
  const [selectedRepo, setSelectedRepo] = useState(null);
  const [repoDetails, setRepoDetails] = useState(null);
  const [commits, setCommits] = useState([]);
  const [loading, setLoading] = useState(false);

  // Function to handle user search
  const handleSearch = async () => {
    try {
      setLoading(true);
      // Fetch user data from the Express backend
      const response = await fetch(
        `http://localhost:3001/api/users?names=${encodeURIComponent(
          JSON.stringify([username])
        )}`
      );
      const data = await response.json();

      if (data && !data.error) {
        // If successful, update user data and fetch repositories
        setUserData(data[0]);
        setError(null);

        // Fetch user repositories from the Express backend
        const reposResponse = await fetch(
          `http://localhost:3001/api/users/${username}/repos`
        );

        if (!reposResponse.ok) {
          throw new Error("Unable to fetch repositories");
        }

        const reposData = await reposResponse.json();
        setRepos(reposData);
        setSelectedRepo(null);
        setRepoDetails(null);
        setCommits([]);
      } else {
        // If there is an error, update error state
        setUserData(null);
        setRepos([]);
        setSelectedRepo(null);
        setRepoDetails(null);
        setCommits([]);
        setError(data.error || "Error searching for user");
      }
    } catch (err) {
      // Handle errors and update error state
      setUserData(null);
      setRepos([]);
      setSelectedRepo(null);
      setRepoDetails(null);
      setCommits([]);
      setError(err.message);
    } finally {
      setLoading(false);
    }
  };

  // Function to handle click on a repository
  const handleRepoClick = async (repoName) => {
    try {
      // Fetch repository details from the Express backend
      const repoDetailsResponse = await fetch(
        `http://localhost:3001/api/users/${username}/repos/${repoName}`
      );

      if (!repoDetailsResponse.ok) {
        throw new Error("Unable to fetch repository details");
      }

      const repoDetailsData = await repoDetailsResponse.json();
      setRepoDetails(repoDetailsData);

      // Fetch the last 5 commits
      const commitsResponse = await fetch(
        `https://api.github.com/repos/${username}/${repoName}/commits?per_page=5`
      );

      if (!commitsResponse.ok) {
        throw new Error("Unable to fetch commits");
      }

      const commitsData = await commitsResponse.json();
      setCommits(commitsData);

      // Set the selected repo for styling purposes
      setSelectedRepo(repoName);
    } catch (err) {
      // Handle errors and update relevant states
      console.error("Error fetching repository details:", err);
      setRepoDetails(null);
      setCommits([]);
    }
  };

  return (
    <div className="container">
      <h1>Github User Search</h1>
      <SearchForm
        username={username}
        setUsername={setUsername}
        handleSearch={handleSearch}
        loading={loading}
      />
      {error && <p className="error">{error}</p>}
      {userData && <UserInfo userData={userData} />}
      {repos.length > 0 && (
        <Repositories
          repos={repos}
          selectedRepo={selectedRepo}
          handleRepoClick={handleRepoClick}
        />
      )}
      {repoDetails && <RepositoryDetails repoDetails={repoDetails} />}
      {commits.length > 0 && <CommitDetails commits={commits} />}
    </div>
  );
}

export default App;
