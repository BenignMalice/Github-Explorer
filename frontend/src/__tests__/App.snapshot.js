// App.test.js
import React from "react";
import renderer from "react-test-renderer";
import App from "../App";

// Snapshot test for the App component
test("App snapshot", () => {
  // Create a renderer for the App component
  const component = renderer.create(<App />);
  // Generate a JSON snapshot of the rendered component
  const tree = component.toJSON();
  // Compare the generated snapshot with the stored snapshot
  expect(tree).toMatchSnapshot();
});
