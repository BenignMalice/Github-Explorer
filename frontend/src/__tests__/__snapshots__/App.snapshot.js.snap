// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<div
  className="container"
>
  <h1>
    Github User Search
  </h1>
  <div>
    <input
      onChange={[Function]}
      placeholder="Enter Github username"
      type="text"
      value=""
    />
    <button
      disabled={false}
      onClick={[Function]}
    >
      Search
    </button>
  </div>
</div>
`;
